compiler/codegen/CCData.hpp

```diff
@@ -0,0 +1,207 @@
+/*******************************************************************************
+ * Copyright (c) 2020, 2020 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef OMR_CCDATA_INCL
+#define OMR_CCDATA_INCL
+
+#include <map>
+#include <string>
+#include <cstddef>
+
+namespace TR { class Monitor; }
+
+namespace OMR
+{
+
+/**
+ * @class CCData
+ *
+ * @brief This class represents a table that can be used to implement a GOT or TOC or a PLT or similar.
+ *
+ * <TODO: A more detailed discussion about the design here.>
+ */
+class CCData
+   {
+   private:
+      /** \typedef data_t Implementation detail. This type represents the units of the table. Typically bytes, but can be some other data type, as long as it can be default-constructed. */
+      typedef uint8_t data_t;
+
+   public:
+      /** \typedef index_t This type represents the indices defined in the public interface of this class. They must behave like integral types. */
+      typedef size_t index_t;
+
+      /** \typedef key_t This type represents the keys defined in the public interface of this class. The constructor is unspecified, use CCData_t::key() to create keys. */
+      typedef std::string key_t;
+
+#if __cpp_static_assert
+      static_assert(sizeof(key_t::value_type) == 1, "Unimplemented key unit size, remaining bytes need to be zeroed to support key units >1.");
```

Comment:
It's more so to help our future selves (the implementers) and make the implicit caveat explicit. I know this code, as written, won't work if keys have padding bytes. If someone wants to change it in the future, for whatever reason, I want to give them a single place to do so and a broken build, rather than a runtime bug, to call attention to the additional work they'd have to do to use their desired key type.